backend/cmd/headlamp_test.go

```diff
@@ -1339,3 +1340,251 @@ func TestProcessTokenProtocol(t *testing.T) {
+
+	resp1, err := httpRequestWithContext(ctx, ts.URL+"/clusters/test/api/v1/resource", "GET")
+	assert.NoError(t, err)
+
+	defer resp1.Body.Close()
+
+	resp2, err := httpRequestWithContext(ctx, ts.URL+"/clusters/test/api/v1/resource", "GET")
+	assert.NoError(t, err)
+
+	defer resp2.Body.Close()
+
+	resp1String, err := stringResponse(resp1)
+	assert.NoError(t, err)
+	resp2String, err := stringResponse(resp2)
+	assert.NoError(t, err)
+
+	assert.Equal(t, expectedResponse, resp1String)
+	assert.Equal(t, "", resp1.Header.Get("X-HEADLAMP-CACHE")) // response is from k8s server, hence X-HEADLAMP-CACHE: ""
+	assert.Equal(t, expectedResponse, resp2String)
+	assert.Equal(t, "true", resp2.Header.Get("X-HEADLAMP-CACHE")) // response is from cache, hence X-HEADLAMP-CACHE: true
+}
+
+// TestCacheMiddleware_AuthErrorResponse test if the user is not authorized
+// to access a resource, CacheMiddleware should return AuthErrorResponse to
+// the client without going to k8 server.
+func TestCacheMiddleware_AuthErrorResponse(t *testing.T) {
+	fakeK8s := setupFakeK8sServer(false)
+	defer fakeK8s.Close()
+
+	c := setupHeadlampConfig(fakeK8s, t.Name())
+
+	proxyHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Use the incoming request's context
+		req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, fakeK8s.URL+r.URL.Path, nil)
+		if err != nil {
+			http.Error(w, "failed to create request", http.StatusInternalServerError)
+			return
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			http.Error(w, "proxy error", http.StatusInternalServerError)
+			return
+		}
+		defer resp.Body.Close()
+
+		_, err = io.Copy(w, resp.Body)
+		assert.NoError(t, err)
+	})
+
+	router := mux.NewRouter()
+	router.PathPrefix("/clusters/{clusterName}/{api:.*}").Handler(
+		CacheMiddleWare(c)(proxyHandler))
+
+	ts := httptest.NewServer(router)
+	defer ts.Close()
+
+	ctx := context.Background()
+
+	expectedResponse := `{"kind":"Status","apiVersion":"v1","metadata":{"resourceVersion":""},` +
+		`"message":"resource is forbidden: User \"system:serviceaccount:default:test\" cannot get resource ` +
+		`\"resource\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"resource"},` +
+		`"code":403}`
+
+	resp1, err := httpRequestWithContext(ctx, ts.URL+"/clusters/test/api/v1/resource", "GET")
+	assert.NoError(t, err)
+
+	defer resp1.Body.Close()
+
+	resp1String, err := stringResponse(resp1)
+	assert.NoError(t, err)
+	assert.Equal(t, expectedResponse, resp1String) // expected authErroResponse to the client.
+	assert.Equal(t, "true", resp1.Header.Get("X-HEADLAMP-CACHE"))
+}
+
+// TestCacheMiddlware_CacheInvalidation test if the request is modifying
+// it should delete the keys, making new fresh request to k8s server and store
+// into the cache if the request is same it should return response from the client.
+func TestCacheMiddlware_CacheInvalidation(t *testing.T) {
```

Comment:
Typo, add "Middlware" should be Middleware. TestCacheMiddlware_CacheInvalidation